ts/src/pro/krakenfutures.ts

```diff
@@ -0,0 +1,1283 @@
+            },
+            'options': {
+                'tradesLimit': 1000,
+                'ordersLimit': 1000,
+                'OHLCVLimit': 1000,
+                'connectionLimit': 100, // https://docs.futures.kraken.com/#websocket-api-websocket-api-introduction-subscriptions-limits
+                'requestLimit': 100, // per second
+                'watchTicker': {
+                    'method': 'ticker', // or ticker_lite
+                },
+                'watchTickers': {
+                    'method': 'ticker', // or ticker_lite
+                },
+                'fetchBalance': {
+                    'type': undefined,
+                },
+            },
+            'streaming': {
+                'keepAlive': 60000,
+            },
+        });
+    }
+
+    async authenticate (params = {}) {
+        /**
+         * @ignore
+         * @method
+         * @description authenticates the user to access private web socket channels
+         * @see https://docs.futures.kraken.com/#websocket-api-public-feeds-challenge
+         * @returns {object} response from exchange
+         */
+        this.checkRequiredCredentials ();
+        // Hash the challenge with the SHA-256 algorithm
+        // Base64-decode your api_secret
+        // Use the result of step 2 to hash the result of step 1 with the HMAC-SHA-512 algorithm
+        // Base64-encode the result of step 3
+        const url = this.urls['api']['ws'];
+        const messageHash = 'challenge';
+        const client = this.client (url);
+        let future = this.safeValue (client.subscriptions, messageHash);
+        if (future === undefined) {
+            const request = {
+                'event': 'challenge',
+                'api_key': this.apiKey,
+            };
+            const message = this.extend (request, params);
+            future = await this.watch (url, messageHash, message);
+            client.subscriptions[messageHash] = future;
+        }
+        return future;
+    }
+
+    async subscribePublic (name: string, symbols: string[], params = {}) {
+        /**
+         * @ignore
+         * @method
+         * @description Connects to a websocket channel
+         * @param {String} name name of the channel
+         * @param {[String]} symbols CCXT market symbols
+         * @param {Object} params extra parameters specific to the krakenfutures api
+         * @returns {Object} data from the websocket stream
+         */
+        await this.loadMarkets ();
+        const url = this.urls['api']['ws'];
+        const subscribe = {
+            'event': 'subscribe',
+            'feed': name,
+        };
+        const marketIds = [ ];
+        let messageHash = name;
+        if (symbols.length === 1) {
+            const symbol = symbols[0];
+            const marketId = this.marketId (symbol);
+            marketIds.push (marketId);
+            messageHash = messageHash + ':' + symbol;
+        }
+        for (let i = 0; i < symbols.length; i++) {
```

Comment:
This does not seem right, if symbols.length === 1 this loop will also run and we will end up with duplicated ids inside marketIds